src/classes/RD_RecurringDonations.cls

```diff
@@ -795,77 +767,391 @@ public class RD_RecurringDonations {
-         
+        list<npe03__Recurring_Donation__c> newlyClosedDonations = new list<npe03__Recurring_Donation__c>();
+
+        List<npe03__Recurring_Donation__c> openEndedRDsToRecalculate = new List<npe03__Recurring_Donation__c>();
+        List<npe03__Recurring_Donation__c> fixedLengthRDsToRecalculate = new List<npe03__Recurring_Donation__c>();
+
         string openlabel = system.label.npe03.RecurringDonationOpenStatus;
         string closedlabel = system.label.npe03.RecurringDonationClosedStatus;        
         
         for (npe03__Recurring_Donation__c r : recurringDonations){
-            npe03__Recurring_Donation__c oldRD = oldRecurringDonations.get(r.id);
-
-            //Check for custom field changes and set flag for update 
-            boolean cfmFieldChanged=false;
-            for (string s : customFieldMappingSettings.keySet()){
-                npe03__Custom_Field_Mapping__c cfm = customFieldMappingSettings.get(s);
-                if(oldRD.get(cfm.npe03__Recurring_Donation_Field__c)!=r.get(cfm.npe03__Recurring_Donation_Field__c)){
-                    cfmFieldChanged=true;
-                }
-            }
-            //if we're closing an open ended donation, modify the existing opps as required by the custom setting
-            if (r.npe03__Open_Ended_Status__c == closedlabel && oldRD.npe03__Open_Ended_Status__c != r.npe03__Open_Ended_Status__c){
-                newlycloseddonations.add(r);        
-            }            
-            
-            //if there's any other change that would require a reevaluation of existing Opps
-            else if (r.npe03__Open_Ended_Status__c != closedlabel){
-            	// for non open-ended rd's, just always re-insert
-            	if (r.npe03__Open_Ended_Status__c != openlabel) {
-            		reinsertopps.add(r);
-            	}
-                else if (r.npe03__Amount__c != oldRD.npe03__Amount__c ||
-                    r.npe03__Recurring_Donation_Campaign__c != oldRD.npe03__Recurring_Donation_Campaign__c ||
-                    r.npe03__Organization__c != oldRD.npe03__Organization__c ||
-                    r.npe03__Next_Payment_Date__c != oldRD.npe03__Next_Payment_Date__c ||
-                    r.Always_Use_Last_Day_Of_Month__c != oldRD.Always_Use_Last_Day_Of_Month__c ||
-                    cfmFieldChanged==true
+
+            Boolean needsRecalc = needsRecalculation(r,oldRecurringDonations.get(r.id));
+
+            if (needsRecalc) {
+                if(r.npe03__Open_Ended_Status__c == closedlabel
+                        //&& oldRD.npe03__Open_Ended_Status__c != r.npe03__Open_Ended_Status__c
                     ){
-                    reevaluateopps.add(r.id);
+                    // closing an open ended donation
+                    // does it actually matter if it's newly closed? shouldn't it go to the runRecurringDonationClosedActions method regardless?
```

Comment:
i'm guessing the logic was that if it was already closed, then there was nothing that would need to get updated.  so for example, if it already was closed, and the user changes the schedule from monthly to yearly, should we do anything?  i'm guessing the intent was no, once it is closed, leave its opps alone.  If you feel we have a case for still updating the RD, lets discuss with PM.